sys/riscv/riscv/trap.c

```diff
@@ -334,6 +335,10 @@ data_abort(struct trapframe *frame, int usermode)
 		ftype = VM_PROT_WRITE;
 	} else if (frame->tf_scause == EXCP_INST_PAGE_FAULT) {
 		ftype = VM_PROT_EXECUTE;
+#if __has_feature(capabilities)
+	} else if (frame->tf_scause == EXCP_STORE_AMO_CAP_PAGE_FAULT) {
+		ftype = VM_PROT_WRITE | VM_PROT_WRITE_CAP;
```

Comment:
Unsure whether this should be both or just VM_PROT_WRITE_CAP; @bsdjhb likely can comment on which is more idiomatic (/ less invasive).

Reply:
Probably just `VM_PROT_WRITE_CAP`.

Reply:
In particular, we don't need to COW the underlying page, etc. or do other things for a write to a non-writable page.  But also my view of `ftype` is it indicates the specific type of fault, and write faults are distinct CPU exceptions from store cap MMU faults.

Reply:
Did we decide to promise that we won't raise `EXCP_STORE_AMO_CAP_PAGE_FAULT` if we could justifiably raise `EXCP_STORE_AMO_PAGE_FAULT` instead?  That seems a little at odds with the RISC-V behaviour of "load side faults of AMOs turn into `STORE_AMO` faults because the store side might fault, too".

Reply:
Hmm, I think for the load side it is clear that the generic load fault should take precedence as the cap fault is data dependent and you can't even see that if you can't read from the page.  For stores you it is perhaps less obvious which fault should take precedence.  My changes over in x86 land actually make this somewhat interesting as there you get a bitmap of flags for page faults and you would very well get a 'cap | write' fault vs a plain 'write' fault (and I probably need to be explicit that 'cap | write' is only for when CW is clear rather than being ambiguous.  I think for `ftype` we still want a single cause to indicate what action to take.  If CHERI-RISC-V doesn't set an explicit precedent, then the pmap code needs to figure out which one it is so that the MI VM code knows what to do as cap write faults require different handling then "can't write at all to the page" faults.

Reply:
So, the reason I went with `VM_PROT_WRITE | VM_PROT_WRITE_CAP` is that `pmap_fault_fixup` can know to look for the `WRITE_CAP` and a very little bit of the VM logic needs to grow to know that `WRITE_CAP` in an `OBJ_NOSTORETAGS` object is fatal, but otherwise everything carries on seeing `VM_PROT_WRITE` asserted (all the tests for that are naturally done by testing that bit specifically).  In the interest of minimizing faults taken, it might be nice if the hardware preferred to signal `EXCP_STORE_AMO_CAP_PAGE_FAULT` for tag-bearing capability stores to absent pages.

Reply:
I think I'm fine with the | for now.